clientconn.go

```diff
@@ -854,176 +850,258 @@ func (ac *addrConn) errorf(format string, a ...interface{}) {
-	for i := ridx; i < len(addrs); i++ {
-		addr := addrs[i]
-		target := transport.TargetInfo{
-			Addr:      addr.Addr,
-			Metadata:  addr.Metadata,
-			Authority: ac.cc.authority,
-		}
-		done := make(chan struct{})
-		onPrefaceReceipt := func() {
-			ac.mu.Lock()
-			close(done)
-			if !ac.backoffDeadline.IsZero() {
-				// If we haven't already started reconnecting to
-				// other backends.
-				// Note, this can happen when writer notices an error
-				// and triggers resetTransport while at the same time
-				// reader receives the preface and invokes this closure.
-				ac.backoffDeadline = time.Time{}
-				ac.connectDeadline = time.Time{}
-				ac.connectRetryNum = 0
-			}
-			ac.mu.Unlock()
-		}
-		// Do not cancel in the success path because of
-		// this issue in Go1.6: https://github.com/golang/go/issues/15078.
-		connectCtx, cancel := context.WithDeadline(ac.ctx, connectDeadline)
-		if channelz.IsOn() {
-			copts.ChannelzParentID = ac.channelzID
-		}
-		newTr, err := transport.NewClientTransport(connectCtx, ac.cc.ctx, target, copts, onPrefaceReceipt)
-		if err != nil {
-			cancel()
-			ac.cc.blockingpicker.updateConnectionError(err)
-			ac.mu.Lock()
-			if ac.state == connectivity.Shutdown {
-				// ac.tearDown(...) has been invoked.
-				ac.mu.Unlock()
-				return false, errConnClosing
+		if err := ac.createTransport(newTrID, backoffIdx, addr, copts, backoffDeadline, onGoAway, onClose); err != nil {
+			// errReadTimeOut indicates that the handshake was not received before
+			// the deadline. We exit here because the transport's reader goroutine will
+			// use onClose to reset the transport.
+			if err == errReadTimedOut {
+				return
 			}
-			ac.mu.Unlock()
-			grpclog.Warningf("grpc: addrConn.createTransport failed to connect to %v. Err :%v. Reconnecting...", addr, err)
-			continue
-		}
-		if ac.dopts.waitForHandshake {
+
+			timer := time.NewTimer(backoffDeadline.Sub(time.Now()))
 			select {
-			case <-done:
-			case <-connectCtx.Done():
-				// Didn't receive server preface, must kill this new transport now.
-				grpclog.Warningf("grpc: addrConn.createTransport failed to receive server preface before deadline.")
-				newTr.Close()
-				continue
+			case <-timer.C:
 			case <-ac.ctx.Done():
+				timer.Stop()
+				return
 			}
+
+			continue
 		}
+
+		return
+	}
+}
+
+var errReadTimedOut = errors.New("read timed out")
+
+// createTransport creates a connection to one of the backends in addrs.
+func (ac *addrConn) createTransport(id int32, backoffNum int, addr resolver.Address, copts transport.ConnectOptions, backoffDeadline time.Time, onGoAway func(transport.GoAwayReason), onClose func()) error {
+	timedOutWaitingForPreface := make(chan struct{})
+	// TODO(deklerk): this is unnecessary. In the reader goroutine, we should be able to signal to onClose that the
+	// deadline was exceeded (we can't use a parameter to t.Close because it would mean changes in too many places)
+	onDeadline := func() {
```

Comment:
Does this need to be passed into the transport?  It seems like you could avoid it by doing (below):

```
	if ac.dopts.waitForHandshake {
		timer := time.NewTimer(connectDeadline.Sub(time.Now()))
		select {
		case <-timer.C: // instead of <-timedOutWaitingForPreface:
			return errReadTimedOut
		case <-prefaceReceived:
			timer.Stop()
		case <-ac.ctx.Done():
			timer.Stop()
			return errConnClosing
		}
	}
```

Reply:
Hmm... we could do this. I don't love two things waiting for the same deadline (this timer + the reader). Following that train of thought, I wondered whether we can move the transport closing behavior out to here and not give the reader a deadline, but I don't think we're guaranteed a non-nil transport here. About to get off the bus, though. Will think on this more this weekend and maybe experiment.

If nothing else, it's not too bad to have two things waiting heh.

Reply:
Looked into it a bit more, and the above is moot since we can kill the reader from clientconn (via t.Close). I've done the needful in https://github.com/grpc/grpc-go/pull/2219/commits/5687feffbf977c00dc0bd792b54927986bffb2a8.